internal/storagenode/logstream/append.go

```diff
@@ -56,23 +57,44 @@ func (at *AppendTask) WaitForCompletion(ctx context.Context) (res []snpb.AppendR
 		return nil, at.err
 	}
 
+	// Append batch requests can succeed partially. In case of partial failure,
+	// failed log entries must be sequential. That is, suffixes of the batch
+	// can be failed to append, whose length can be from zero to the full size.
+	hasNonContextErr := false
 	res = make([]snpb.AppendResult, at.dataBatchLen)
 	for i := range at.apc.awgs {
 		cerr := at.apc.awgs[i].wait(ctx)
 		if cerr != nil {
+			if !hasNonContextErr {
+				hasNonContextErr = !errors.Is(cerr, context.Canceled) && !errors.Is(cerr, context.DeadlineExceeded)
+			}
+
 			res[i].Error = cerr.Error()
 			if err == nil {
 				err = cerr
 			}
 			continue
 		}
-		if err != nil {
+
+		// It has not failed yet.
+		if err == nil {
+			res[i].Meta.TopicID = at.lse.tpid
+			res[i].Meta.LogStreamID = at.lse.lsid
+			res[i].Meta.GLSN = at.apc.awgs[i].glsn
+			res[i].Meta.LLSN = at.apc.awgs[i].llsn
+			at.apc.awgs[i].release()
+			continue
+		}
+
+		// It panics when the batch's success and failure are interleaved.
+		// However, context errors caused by clients can be ignored since
+		// the batch can succeed in the storage node, although clients
+		// canceled it.
+		// Once the codebase stabilizes, it is planned to be removed.
+		if hasNonContextErr {
```

Comment:
Is this happening when a `context.Canceled` error occurred in the past, but this time the append succeeded(cerr == nil)?

Reply:
I just checked it again, it may happen currently.
The situation seems to be avoided by checking the context error by using the `if` clause rather than `select`.

Although `ctx.Done` is closed, `wwg.ch` can also be closed.

https://github.com/kakao/varlog/blob/0acedd9afc9d44c5b4df2b05623e2fdcb5d14515/internal/storagenode/logstream/append_waitgroup.go#L38-L45

Reply:
I added codes that check the context error. It avoids that cerr in `cerr := at.apc.awgs[i].wait(ctx)` is nil after context error.

https://github.com/kakao/varlog/blob/3bec88e250a9b2671df52622adaa29e0624edb4f/internal/storagenode/logstream/append_waitgroup.go#L145-L153

Regardless of the addition of the codes, panic must not occur.